In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import pylab as pl
import numpy as np

In [2]:
df_raw = pd.read_csv(r'C:\Users\animo\Documents\GA\DAT-BOS-16-master\projects\unit-projects\project-3\assets\admissions.csv')
df = df_raw.dropna()
print df.head()

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


Part 1. Frequency Table of our variables

In [9]:
df_raw.describe()

admit         gre        gpa    prestige
count  400.000000  398.000000  398.00000  399.000000
mean     0.317500  588.040201    3.39093    2.486216
std      0.466087  115.628513    0.38063    0.945333
min      0.000000  220.000000    2.26000    1.000000
25%      0.000000  520.000000    3.13000    2.000000
50%      0.000000  580.000000    3.39500    2.000000
75%      1.000000  660.000000    3.67000    3.000000
max      1.000000  800.000000    4.00000    4.000000

In [11]:
#for prestige and whether or not a person was admitted
print pd.crosstab(df['admit'], df['prestige'], rownames=['admit'])

prestige  1.0  2.0  3.0  4.0
admit                       
0          28   95   93   55
1          33   53   28   12


Part 2. Return of dummy variables. Create class or dummy variables for 'Prestige'.

In [13]:
pd.get_dummies(df['prestige'], prefix='prestige')

prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0               0             0             1             0
1               0             0             1             0
2               1             0             0             0
3               0             0             0             1
4               0             0             0             1
5               0             1             0             0
6               1             0             0             0
7               0             1             0             0
8               0             0             1             0
9               0             1             0             0
10              0             0             0             1
11              1             0             0             0
12              1             0             0             0
13              0             1             0             0
14              1             0             0             0
15              0             0             1             0
16              0             0             0             1
17              0             0             1             0
18              0             1             0             0
19              1             0             0             0
20              0             0             1             0
21              0             1             0             0
22              0             0             0             1
23              0             0             0             1
24              0             1             0             0
25              1             0             0             0
26              1             0             0             0
27              0             0             0             1
28              0             1             0             0
29              1             0             0             0
..            ...           ...           ...           ...
370             0             1             0             0
371             0             0             1             0
372             1             0             0             0
373             1             0             0             0
374             0             1             0             0
375             0             0             0             1
376             0             1             0             0
377             0             1             0             0
378             0             0             1             0
379             0             1             0             0
380             0             1             0             0
381             0             1             0             0
382             0             1             0             0
383             1             0             0             0
384             0             1             0             0
385             1             0             0             0
386             0             1             0             0
387             0             1             0             0
388             0             1             0             0
389             0             1             0             0
390             0             1             0             0
391             0             1             0             0
392             0             0             1             0
393             0             1             0             0
394             0             0             1             0
395             0             1             0             0
396             0             0             1             0
397             0             1             0             0
398             0             1             0             0
399             0             0             1             0

[397 rows x 4 columns]

In [14]:
dummy_ranks = pd.get_dummies(df['prestige'], prefix='prestige')
print dummy_ranks.head()

   prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0             0             0             1             0
1             0             0             1             0
2             1             0             0             0
3             0             0             0             1
4             0             0             0             1


We'll only need one dummy variable class when modelling.

Hand calculating odds ratios

In [34]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_1.0':])
print handCalc.head()

   admit    gre   gpa  prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61             0             0             1             0
1      1  660.0  3.67             0             0             1             0
2      1  800.0  4.00             1             0             0             0
3      1  640.0  3.19             0             0             0             1
4      0  520.0  2.93             0             0             0             1


In [25]:
handCalc = df[cols_to_keep].join(dummy_ranks.ix[:,'prestige_1.0'])
print handCalc

     admit    gre   gpa  prestige_1.0
0        0  380.0  3.61             0
1        1  660.0  3.67             0
2        1  800.0  4.00             1
3        1  640.0  3.19             0
4        0  520.0  2.93             0
5        1  760.0  3.00             0
6        1  560.0  2.98             1
7        0  400.0  3.08             0
8        1  540.0  3.39             0
9        0  700.0  3.92             0
10       0  800.0  4.00             0
11       0  440.0  3.22             1
12       1  760.0  4.00             1
13       0  700.0  3.08             0
14       1  700.0  4.00             1
15       0  480.0  3.44             0
16       0  780.0  3.87             0
17       0  360.0  2.56             0
18       0  800.0  3.75             0
19       1  540.0  3.81             1
20       0  500.0  3.17             0
21       1  660.0  3.63             0
22       0  600.0  2.82             0
23       0  680.0  3.19             0
24       1  760.0  3.35             0
25       1  

In [30]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(dummy_ranks.ix[:,'prestige_1.0'])
print handCalc

     admit    gre   gpa  prestige_1.0
0        0  380.0  3.61             0
1        1  660.0  3.67             0
2        1  800.0  4.00             1
3        1  640.0  3.19             0
4        0  520.0  2.93             0
5        1  760.0  3.00             0
6        1  560.0  2.98             1
7        0  400.0  3.08             0
8        1  540.0  3.39             0
9        0  700.0  3.92             0
10       0  800.0  4.00             0
11       0  440.0  3.22             1
12       1  760.0  4.00             1
13       0  700.0  3.08             0
14       1  700.0  4.00             1
15       0  480.0  3.44             0
16       0  780.0  3.87             0
17       0  360.0  2.56             0
18       0  800.0  3.75             0
19       1  540.0  3.81             1
20       0  500.0  3.17             0
21       1  660.0  3.63             0
22       0  600.0  2.82             0
23       0  680.0  3.19             0
24       1  760.0  3.35             0
25       1  

Assigning dummies to prestige_1 will help avoid multicollinearity.

In [33]:
print pd.crosstab(df['admit'], df['prestige'], rownames=['admit'])
print pd

prestige  1.0  2.0  3.0  4.0
admit                       
0          28   95   93   55
1          33   53   28   12
<module 'pandas' from 'C:\Users\animo\Anaconda2\envs\p27\lib\site-packages\pandas\__init__.pyc'>


In [36]:
print pd.crosstab(handCalc['admit'], handCalc['prestige_1.0'], rownames=['admit'])
print pd

prestige_1.0    0   1
admit                
0             243  28
1              93  33
<module 'pandas' from 'C:\Users\animo\Anaconda2\envs\p27\lib\site-packages\pandas\__init__.pyc'>


Odds of being admitted to grad school if you attended a #1 ranked college:

In [37]:
odds_exposed = 33.0 / 28
print odds_exposed

1.17857142857


Odds of being admitted if you didn't attend a #1 ranked college:

In [38]:
odds_unexp = 93.0 / 243.0
print odds_unexp

0.382716049383


Odds ratio

In [39]:
OR = odds_exposed / odds_unexp
print OR

3.07949308756


Students who attended a top ranked college are 3.07 more likely to be admitted to graduate school than those students who did not attend a top ranked college.

In [40]:
print pd.crosstab(handCalc['admit'], handCalc['prestige_4.0'], rownames=['admit'])
print pd

prestige_4.0    0   1
admit                
0             216  55
1             114  12
<module 'pandas' from 'C:\Users\animo\Anaconda2\envs\p27\lib\site-packages\pandas\__init__.pyc'>


In [42]:
OR = (12.0*216.0)/(114.0*55.0)
print OR

0.413397129187


In [44]:
odds_exp = 12.0/55.0
print odds_exp
odds_unexp= 114.0 / 216.0
print odds_unexp


0.218181818182
0.527777777778


In [45]:
OR = odds_exp / odds_unexp
print OR

0.413397129187


The odds that a student who attended a low ranked college is admitted to graduate school are .41 times the odds of students who did not attend a low ranked college. 

Part 4. Analysis

In [48]:
cols_to_keep = ['admit', 'gre', 'gpa']
data = df[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_2.0':])
print data.head()

   admit    gre   gpa  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61             0             1             0
1      1  660.0  3.67             0             1             0
2      1  800.0  4.00             0             0             0
3      1  640.0  3.19             0             0             1
4      0  520.0  2.93             0             0             1


In [49]:
data['intercept']=1.0

In [51]:
train_cols = data.columns[1:]

Fit the model

In [52]:
logit = sm.Logit(data['admit'], data[train_cols])
result = logit.fit()

Optimization terminated successfully.
         Current function value: 0.573854
         Iterations 6


In [53]:
print result.summary()

                           Logit Regression Results                           
Dep. Variable:                  admit   No. Observations:                  397
Model:                          Logit   Df Residuals:                      391
Method:                           MLE   Df Model:                            5
Date:                Thu, 04 May 2017   Pseudo R-squ.:                 0.08166
Time:                        13:17:36   Log-Likelihood:                -227.82
converged:                       True   LL-Null:                       -248.08
                                        LLR p-value:                 1.176e-07
                   coef    std err          z      P>|z|      [95.0% Conf. Int.]
--------------------------------------------------------------------------------
gre              0.0022      0.001      2.028      0.043      7.44e-05     0.004
gpa              0.7793      0.333      2.344      0.019         0.128     1.431
prestige_2.0    -0.6801      0.317     -2.14

In [56]:
print np.exp(result.params)

gre             1.002221
gpa             2.180027
prestige_2.0    0.506548
prestige_3.0    0.262192
prestige_4.0    0.211525
intercept       0.020716
dtype: float64


In [57]:
params = result.params
conf = result.conf_int()
conf['OR'] = params
conf.columns = ['2.50%', '97.50%', 'OR']
print np.exp(conf)

                 2.50%    97.50%        OR
gre           1.000074  1.004372  1.002221
gpa           1.136120  4.183113  2.180027
prestige_2.0  0.272168  0.942767  0.506548
prestige_3.0  0.133377  0.515419  0.262192
prestige_4.0  0.093329  0.479411  0.211525
intercept     0.002207  0.194440  0.020716


Those students who attended a Prestige 2 undergraduate institution had .50 the odds of being admitted to grad school versus those students who attended a top tier undergraduate school.

Those students with a higher GPA had 2.18 odds of being admitted to a graduate school.

In [83]:
def cartesian(arrays, out=None):

    cartesian(([1,2,3], [4,5], [5,6]))
    
    array = [np.arrays(x) for x in arrays]
    dtype = arrays[0].dtype
    
n = np.prod([x.size for x in arrays])
if out is None:
    out = np.zeros([n, len(arrays)], dtype=dtype)
    
m = n / arrays[0].size
out[:,0] = np.repeat(arrays[0], m)
if arrays[1:]:
    cartesian(arrays[1:], out=out[0:m,1:])
    for j in xrange(1, arrays[0].size):
        out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

NameError: name 'arrays' is not defined

In [74]:
gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)
print gres

[ 220.          284.44444444  348.88888889  413.33333333  477.77777778
  542.22222222  606.66666667  671.11111111  735.55555556  800.        ]


In [76]:
gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)
print gpas

[ 2.26        2.45333333  2.64666667  2.84        3.03333333  3.22666667
  3.42        3.61333333  3.80666667  4.        ]


Recreate the dummy variables

In [87]:
combos = pd.DataFrame(cartesian([gres, gpas, [1,2,3,4], [1.]]))

combos.columns = ['gre', 'gpa', 'prestige', 'intercept']
dummy_ranks = pd.get_dummies(combos['prestige'], prefix='prestige')
dummy_ranks.columns = ['prestige_1.0', 'prestige_2.0,', 'prestige_3.0', 'prestige_4.0']

cols_to_keep = ['gre', 'gpa', 'prestige', 'intercept']
combos = combos[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_2.0':])

RuntimeError: maximum recursion depth exceeded

Make predictions on the enumerated dataset

In [94]:
combos = pd.DataFrame(cartesian([gres, gpas, [1,2,3,4], [1.]]))
combos['admit_pred'] = result.predict(combos[train_cols])

print combos.tail()
len(combos)

RuntimeError: maximum recursion depth exceeded

Given the same GPA and GRE scores, students attending a top ranked undergraduate institution have a 73% probability of being admitted to graduate school. Those attending a tier two or tier three undergraduate school have a 58% and 41% likelihood, respectively. Those students attending tier four undergraduate schools have a 37% probability of being admitted to graduate school.  